### PR TITLE
MAINT, CI: cibuildwheel support

### PR DESCRIPTION
**Pull request by shanedsnyder**
This PR includes updates for supporting `cibuildwheel`:

- `cibuildwheel` configuration in `pyproject.toml` which works for my local Linux builds over Docker
- first cut at GitHub workflow for automatically building wheels on updates to code that affects PyDarshan
- change `setup.py` argument `--with-extension` to env variable `PYDARSHAN_BUILD_EXT` 

The last change was needed because I couldn't find any other way to get user arguments to `setup.py` from `cibuildwheel`. There seems to be lack of general support in `cibuildwheel` for user-specified options to `setup.py`, as well as issues with `pip` supporting the `--build-option` flag we had been using -- see discussion at https://github.com/pypa/cibuildwheel/issues/268. The issue with `--build-option` is what I think was leading to recurring issues with our existing wheel building setup, with that option being silently ignored and the extension ultimately not being built.

WIP tag while I verify the GitHub actions are working as my local test have.

### CODE REVIEW

File: .github/workflows/build_wheels.yml
Comment (tylerjereddy) [inline .github/workflows/build_wheels.yml:15]: I just wanted to make a suggestion here based on something I saw in another project:

> Numpy currently uses `cibuildwheel <https://https://cibuildwheel.readthedocs.io/en/stable/>`_
in order to build wheels through continuous integration services. To save resources, the
cibuildwheel wheel builders are not run by default on every single PR or commit to main.

> If you would like to test that your pull request do not break the wheel builders,
you may either append ``[wheel build]`` to the end of the commit message of the commit...

and then the CI config has something like this:

> .github/workflows/wheels.yml:      contains(needs.get_commit_message.outputs.message, '[wheel build]')

maybe you'll want to defer that to another issue, but wanted to mention that

File: darshan-util/pydarshan/pyproject.toml
Comment (tylerjereddy) [inline darshan-util/pydarshan/pyproject.toml:27]: It may be worth a comment or issue or something like that to remind us to make the logs repo available for testing here eventually, so that wheels get a thorough testing treatment before release.

We can't just add the project name here, as you likely realized per https://github.com/darshan-hpc/darshan-logs/issues/15

File: .github/workflows/build_wheels.yml
Comment (shanedsnyder) [reply .github/workflows/build_wheels.yml:15]: Yeah, I like that idea and planned on asking for feedback for that specifically, because the wheel building process takes a _long_ time (especially for PyPy builds for whatever reason). Building them on demand (maybe on big changes and at release time) makes a of sense. I'll try to address that in this PR once I'm sure things are running properly otherwise. Thanks for the pointer.

File: darshan-util/pydarshan/pyproject.toml
Comment (shanedsnyder) [reply darshan-util/pydarshan/pyproject.toml:27]: Good idea. You could probably make `cibuildwheel` just clone the repo and install it from source, but we should just make a PyPI package probably. I'll make an issue to track that.

**Comment by shanedsnyder** [comment]
This is kind of working, but still some things left to resolve:

- PyPy 3.9 builds for macosx are failing due to not finding openblas/lapack scipy dependencies (despite brew installing them)
- PyPy 3.9 builds for manylinux were failing at some point, causing Python to abort (segfault?) in some obscure callstack
- restrict wheel builds to on-demand along lines mentioned by Tyler in previous comment
- open issue to remind us to integrate darshan-logs package once it's available on PyPI

There's also something that's slightly confusing to me, in that the testing phase is taking a really long time (30+ minutes on PyPy builds), presumably because it looks like most (all?) of the dependencies are being built from source (which is also causing issue 1 above). Does that make sense? Not sure why they can't just be installed from binary packages on the index just as I would do if running manual tests locally.

File: darshan-util/pydarshan/pyproject.toml
Comment (jakobluettgau) [inline darshan-util/pydarshan/pyproject.toml:22]: From the example: https://cibuildwheel.readthedocs.io/en/stable/options/#configuration-file
```
[tool.cibuildwheel.linux]
repair-wheel-command = "auditwheel repair -w {dest_dir} {wheel}"
```

And more info here:
https://cibuildwheel.readthedocs.io/en/stable/options/#repair-wheel-command

File: darshan-util/pydarshan/pyproject.toml
Comment (jakobluettgau) [inline darshan-util/pydarshan/pyproject.toml:33]: From the example: https://cibuildwheel.readthedocs.io/en/stable/options/#configuration-file
```
[tool.cibuildwheel.macos]
repair-wheel-command = [
  "delocate-listdeps {wheel}",
  "delocate-wheel --require-archs {delocate_archs} -w {dest_dir} {wheel}",
]
```

And more info here:
https://cibuildwheel.readthedocs.io/en/stable/options/#repair-wheel-command

**Comment by jakobluettgau** [comment]
One other thing that occurred to me is that, while the submodules initialized and the darshan-utils shared object is build with them enabled, the python files from those are not copied over yet into the pydarshan/darahsn/backend directory, right?

`cd ./darshan-util/pydarshan && make add-modules`

File: darshan-util/pydarshan/pyproject.toml
Comment (shanedsnyder) [reply darshan-util/pydarshan/pyproject.toml:22]: Thanks for catching that, I'll have a look at this and below and try to add them to the pyprojec.toml

**Comment by tylerjereddy** [comment]
> There's also something that's slightly confusing to me, in that the testing phase is taking a really long time (30+ minutes on PyPy builds), presumably because it looks like most (all?) of the dependencies are being built from source

I'm not sure why we'd even want to offer PyPy wheels? This is done at the base of the ecosystem only in exceptional circumstances where there is a strong evidence of a performance benefit. Have we really done detailed studies justifying the usage of the JIT-compiled version (PyPy) of pydarshan vs. conventional CPython wheels? We're not testing with PyPy in our CI normally, and we don't do benchmarks with PyPy at the moment either.

At the moment, it looks like wheel builds are confined to PyPy (`build = "pp39*"`) in this PR, but I'd suggest just avoiding PyPy altogether and using the conventional interpreted CPython implementation of Python (`cp*` type builds--an exclusion list like this for example: `skip = "pp* *-manylinux_i686 *_ppc64le *_s390x *-musllinux*"`; I did something like that for my early testing over in https://github.com/scipy/scipy/pull/15925).

As to the original question--yes, it makes sense those builds take forever and have problems because many packages like SciPy won't provide wheels--we found that PyPy support was just too great of a burden (it is an entirely different implementation of Python itself), even with a PyPy dev in our community.

**Comment by shanedsnyder** [comment]
Okay, I think I've addressed all comments here and am content with where things are at. My most recent commit updates our PyDarshan release checklist to make clear how we could take wheels generated from actions and upload them to PyPI.

I dropped the WIP tag, but I'll leave it to you guys whether there's more to address here. If you do approve the PR and merge it, please do make sure to squash into a single commit -- a lot of these commits were just trying to get CI to run and things like that, so don't want all the noise in the git history.

**Comment by tylerjereddy** [comment]
When I run with this branch locally: `cibuildwheel --platform linux`

I get an error:

```
...
Installed:
  libjpeg-turbo.x86_64 0:1.2.90-8.el7 libjpeg-turbo-devel.x86_64 0:1.2.90-8.el7

Complete!

                                                             ✕ 32.95s
Error: Command ['sh', '-c', 'yum install -y blas-devel lapack-devel && yum install -y libxslt-devel libxml2-devel && yum install -y libjpeg libjpeg-devel && git submodule update --init && ./prepare.sh && ./configure --disable-darshan-runtime --enable-apxc-mod --enable-apmpi-mod && make install'] failed with code 128. 
```

**Comment by tylerjereddy** [comment]
A bit more info here, `git submodule update --init` is the command that causes that particular issue. A bit of a guess, but maybe there is a bit of magic happening with the `package-dir` variable in CI vs. what happens when I run from within `pydarshan` folder locally? Different effective working directory/context or something?

**Comment by shanedsnyder** [comment]
Oh, maybe it's this: you need to run cibuildwheel from the top-level directory in the repo so it can properly build darshan-util. I've been running it like:

`cibuildwheel --platform linux darshan-util/pydarshan`

**Comment by tylerjereddy** [comment]
Ok, that works, it may be worth documenting somewhere.

```
5 wheels produced in 16 minutes:
  darshan-3.4.0.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl   856 kB
  darshan-3.4.0.0-cp36-cp36m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl    856 kB
  darshan-3.4.0.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl    856 kB
  darshan-3.4.0.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl     856 kB
  darshan-3.4.0.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl     855 kB

```

**Comment by tylerjereddy** [comment]
Ok, I think this can go in now. It is quite similar to some of the upstream configs, like the one used by NumPy, and is a big step forward for our wheel/release process.

gh-744 covers my main concern for a follow-up; one other question is whether we really want `path` directives on the wheel build process? It could be that the release manager will just want to build wheels without changing files at a particular path. I think the manual trigger may work in that case though, not sure.

**Review by tylerjereddy — APPROVED** [review]

**Comment by tylerjereddy** [comment]
squash merged in

**Pull request merged**
